src/main/java/org/terasology/buildings/BuildingUpgradeSystem.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    @ReceiveEvent
+    public void onActivateEvent(ActivateEvent event, EntityRef entityRef) {
+        logger.info("event received activate");
+        EntityRef player = event.getInstigator();
+        CharacterHeldItemComponent heldItemComponent = player.getComponent(CharacterHeldItemComponent.class);
+
+        EntityRef selectedItem = heldItemComponent.selectedItem;
+        DisplayNameComponent displayNameComponent = selectedItem.getComponent(DisplayNameComponent.class);
+
+        if (displayNameComponent != null && !displayNameComponent.name.equals(Constants.UPGRADE_TOOL_NAME)) {
+            return;
+        }
+
+        Vector3f hitBlockPos = event.getTargetLocation();
+
+        checkIfPartOfBuilding(hitBlockPos, player);
+    }
+
+    /**
+     * Checks if the block activated using the upgrade tool is part of a building
+     * @param blockPos The position of the block to be checked
+     * @param player The entity which sends the event
+     */
+    private void checkIfPartOfBuilding(Vector3f blockPos, EntityRef player) {
+        HoldingComponent holdingComponent = player.getComponent(HoldingComponent.class);
+        List<EntityRef> buildings = holdingComponent.constructedBuildings;
+
+        for (EntityRef building : buildings) {
+            ConstructedBuildingComponent buildingComponent = building.getComponent(ConstructedBuildingComponent.class);
+            List<Region3i> buildingRegions = buildingComponent.boundingRegions;
+
+            for (Region3i buildingRegion : buildingRegions) {
+                Vector3i target = new Vector3i(blockPos);
+                if (buildingRegion.encompasses(target)) {
+                    logger.info("hit building" + buildingComponent.buildingType);
+                    this.buildingToUpgrade = building;
+                    building.send(new OpenUpgradeScreenEvent());
+                }
+            }
+        }
+    }
+
+    public EntityRef getBuildingToUpgrade() {
+        return buildingToUpgrade;
+    }
+
+    /**
+     * Receives the {@link UpgradeBuildingEvent} triggered by the Upgrade Button on the BuildingUpgradeScreen and adds a
+     * upgrade task to the holding
+     * @param upgradeBuildingEvent The event received
+     */
+    @ReceiveEvent
+    public void onReceiveBuildingUpgradeEvent(UpgradeBuildingEvent upgradeBuildingEvent, EntityRef player) {
+        buildingToUpgrade.send(new CloseUpgradeScreenEvent());
+        logger.info("adding building upgrade task");
+
+        ConstructedBuildingComponent buildingComponent = buildingToUpgrade.getComponent(ConstructedBuildingComponent.class);
+
+        TaskComponent taskComponent = new TaskComponent();
+        taskComponent.assignedTaskType = AssignedTaskType.Upgrade;
+        // TODO: Assign a random region or a region based on blocks to be upgraded
+        taskComponent.taskRegion = buildingComponent.boundingRegions.get(86);
```

Comment:
What's "86"?   Should be a constant somewhere which describes what this number represents.

Reply:
`86` is the region around a chair in the diner which is accessible by the Oreon and it does not get stuck anywhere trying to reach it. So I used this for testing purpose, this way the Oreon does not abandon an assigned `eat` task.

Reply:
So you probably should use Constants.DINER_CHAIR_LOCATION instead of 86.